soroban-env-host/src/host.rs

```diff
@@ -741,41 +744,69 @@ impl Host {
     }
 
     // Notes on metering: this is covered by the called components.
-    pub fn create_contract_with_id(
+    fn create_contract_with_id(
         &self,
-        contract: ScContractCode,
-        id_obj: Object,
+        contract_id: Object,
+        contract_code: ScContractCode,
     ) -> Result<(), HostError> {
-        let new_contract_id = self.hash_from_obj_input("id_obj", id_obj)?;
+        let new_contract_id = self.hash_from_obj_input("id_obj", contract_id)?;
         let storage_key =
-            self.contract_code_ledger_key(new_contract_id.metered_clone(&self.0.budget)?);
+            self.contract_code_handle_ledger_key(new_contract_id.metered_clone(&self.0.budget)?);
         if self.0.storage.borrow_mut().has(&storage_key)? {
             return Err(self.err_general("Contract already exists"));
         }
-        self.store_contract_code(contract, new_contract_id, &storage_key)?;
+        // Make sure the contract code exists. With immutable contracts and
+        // without this check it would be possible to accidentally create a
+        // contract that never may be invoked (just by providing a bad hash).
+        if let ScContractCode::WasmRef(wasm_hash) = &contract_code {
+            let wasm_storage_key =
+                self.contract_code_ledger_key(wasm_hash.metered_clone(&self.0.budget)?);
+            if !self.0.storage.borrow_mut().has(&wasm_storage_key)? {
+                return Err(self.err_general("Contract code was not installed"));
+            }
+        }
+        self.store_contract_code(contract_code, new_contract_id, &storage_key)?;
```

Comment:
Hmm source is confusing to me because I expect it to contain the source code, which isn't true here. Maybe you should leave it and open a cleanup issue? What do you think?

Reply:
The source is now consistently consistently refers to the implementation (either wasm or builtin). That's basically the 'class name/type' of the contract in OOP analogy. I realize this may be a bit confusing, but OTOH not sure what would be a better/less confusing name. Since changes would need to be done in XDR, let's maybe brainstorm this a bit and cleanup later.

Reply:
Opened https://github.com/stellar/rs-soroban-env/issues/575